### PR TITLE
Stress test

### PR DESCRIPTION
**Pull request by LittleJennie**
- Stress testing with Artillery and K6
- Optimize server with Redis

### CODE REVIEW

File: newrelic.js [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+ */
+exports.config = {
+  /**
+   * Array of application names.
+   */
+  app_name: ['Sigsa7_reviews'],
+  /**
+   * Your New Relic license key.
+   */
+  license_key: '5cbd8b69ee96335a1c9eacca7da172e4ca813663',
```
Comment (kwang004) [inline newrelic.js]: @LittleJennie do you want to add this to your gitignore

Comment (LittleJennie) [reply newrelic.js]: I think this is ignored.

File: .gitignore [hunk 2]
```diff
@@ -7,4 +7,7 @@ server/db/postgres_restaurants.csv
 server/db/postgres_reviews.csv
 server/db/mongoRestaurant.csv
 server/db/mongoReviews.csv
-server/db/mongoUsers.csv
\ No newline at end of file
+server/db/mongoUsers.csv
+newrelic.js
+newrelic_agent.log
+dump.rdb
```
Comment (NickGod) [inline .gitignore]: new line at the end? nit

File: k6StressTest.js [hunk 3]
```diff
@@ -0,0 +1,36 @@
+import http from 'k6/http';
+import { sleep } from 'k6';
+import faker from 'https://cdnjs.cloudflare.com/ajax/libs/Faker/3.1.0/faker.min.js';
```
Comment (NickGod) [inline k6StressTest.js]: why can't we import this as a normal package?

File: server/Models/getRestaurantInfo.js [hunk 4]
```diff
@@ -1,66 +1,74 @@
...
+const redis = require('../db/redisConnection');
 
-const getRestaurantInfo = (req, res) => {
+const getRestaurantInfo = async (req, res) => {
   const { restaurantId } = req.params;
 
-  const { sort, keyword, star } = req.body;
+  const { sort, keywords, star } = req.body;
   let orderByStr = '';
   let selectStr = orderByStr;
```
Comment (NickGod) [inline server/Models/getRestaurantInfo.js]: what is the point of assigning orderByStr to selectStr?

File: server/Models/getRestaurantInfo.js [hunk 5]
```diff
@@ -1,66 +1,74 @@
...
   let keywordStr = '';
   let starStr = '';
 
   if (sort === 'highest rating') {
-    orderByStr = 'CAST((reviews.foodrating+reviews.servicerating+reviews.ambiencerating+reviews.valuerating) AS FLOAT)/4';
+    orderByStr = ', CAST((reviews.foodrating+reviews.servicerating+reviews.ambiencerating+reviews.valuerating) AS FLOAT)/4  AS avg_overal_from_reviews';
     sortDirection = ' DESC,';
   }
 
   if (sort === 'lowest rating') {
```
Comment (NickGod) [inline server/Models/getRestaurantInfo.js]: probably do else if? nit

File: server/Models/getRestaurantInfo.js [hunk 6]
```diff
@@ -1,66 +1,74 @@
...
-    orderByStr = 'CAST((reviews.foodrating+reviews.servicerating+reviews.ambiencerating+reviews.valuerating) AS FLOAT)/4';
+    orderByStr = ', CAST((reviews.foodrating+reviews.servicerating+reviews.ambiencerating+reviews.valuerating) AS FLOAT)/4 AS avg_overal_from_reviews';
     sortDirection = ' ASC,';
   }
 
-  if (keyword !== undefined) {
-    for (let i = 0; i < keyword.length; i ++) {
-      keywordStr += `AND reviews.reviewText LIKE '%${keyword[i]}%' `;
+  if (keywords.length !== 0) {
```
Comment (NickGod) [inline server/Models/getRestaurantInfo.js]: if (keywords && keywords.length !== 0)

how do you guarantee keywords won't be undefined?

File: server/Models/getRestaurantInfo.js [hunk 7]
```diff
@@ -1,66 +1,74 @@
...
 
   if (sort === 'highest rating') {
-    orderByStr = 'CAST((reviews.foodrating+reviews.servicerating+reviews.ambiencerating+reviews.valuerating) AS FLOAT)/4';
+    orderByStr = ', CAST((reviews.foodrating+reviews.servicerating+reviews.ambiencerating+reviews.valuerating) AS FLOAT)/4  AS avg_overal_from_reviews';
     sortDirection = ' DESC,';
   }
 
   if (sort === 'lowest rating') {
-    orderByStr = 'CAST((reviews.foodrating+reviews.servicerating+reviews.ambiencerating+reviews.valuerating) AS FLOAT)/4';
+    orderByStr = ', CAST((reviews.foodrating+reviews.servicerating+reviews.ambiencerating+reviews.valuerating) AS FLOAT)/4 AS avg_overal_from_reviews';
```
Comment (NickGod) [inline server/Models/getRestaurantInfo.js]: don't like prepending comma here

File: server/Models/getRestaurantInfo.js [hunk 8]
```diff
@@ -1,66 +1,74 @@
...
+    if (result) {
+      const jsonResult = JSON.parse(result);
+      return res.status(200).json(jsonResult.rows);
+
     } else {
-      res.status(200).json(data.rows);
+      try {
+        const data = await pg.query(query);
+        const stringfyData = JSON.stringify(data);
+        redis.setex(query.text, 1800, stringfyData);
```
Comment (NickGod) [inline server/Models/getRestaurantInfo.js]: no need to make another variable, can just do
redis.setex(query.text, 1800, JSON.stringify(data))

File: stress_test.yml [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+    flow:
+      - function: "generateGetRequestData"
+      - loop:
+        - get:
+            url: "/{{ restaurantId }}/reviews"
+            json:
+              sort: "{{ sort }}"
+              star: "{{ star }}"
+              keywords: "{{ keywords }}"
+        count: 100
```
Comment (NickGod) [inline stress_test.yml]: new line after?

File: server/Models/getRestaurantInfo.js (hunk 7 above)
Comment (NickGod) [reply server/Models/getRestaurantInfo.js]: can ignore

File: k6StressTest.js (hunk 3 above)
Comment (LittleJennie) [reply k6StressTest.js]: import does not support node packages installed in app.

File: server/Models/getRestaurantInfo.js (hunk 6 above)
Comment (LittleJennie) [reply server/Models/getRestaurantInfo.js]: I was thinking to restrict `keywords` to be sent as an empty array by default from the client. But you are right, I should define the condition where `keywords` is undefined.

**Pull request merged**